File: Assets/scripts/damageableComponent.js
#pragma strict
// allows an object to be damaged. Works with resource passing modules on the object if there are any.
var health : float = 100;
var maxHealth : float = 100;
//var destroyOnZero : boolean = false; // set to true to have the object destroyed when it reaches 0 hp.
var resourceScript : resourcePasser;
var generatorScript : resourceGenerator;
function Start () {
	resourceScript = gameObject.GetComponent("resourcePasser");
	generatorScript = gameObject.GetComponent("resourceGenerator");
}


function TakeDamage (amount : float) {
	health -= amount;
	health = CheckBounds(health, 0, maxHealth);
	SetResourceRate();
	SetColor();
}

function SetHealth(amount : float) {
	health = amount;
	
	health = CheckBounds(health, 0, maxHealth);
	SetResourceRate();
	SetColor();
}

function Repair (amount : float) {
	health += amount;
	health = CheckBounds(health, 0, maxHealth);
	SetResourceRate();
	SetColor();
}

function SetResourceRate () {
	// could also use sendmessage here
	if (resourceScript != null) {
		var ratePercentage = resourceScript.maxRate * (health / maxHealth);
		resourceScript.SetRate(ratePercentage);
	}
	
	if (generatorScript != null) {
		var generatorPercentage = generatorScript.maxRate * (health / maxHealth);
		generatorScript.SetRate(generatorPercentage);
	}
}

function SetColor () {
	var healthPercentage = health / maxHealth;
	gameObject.GetComponent.<Renderer>().material.SetColor("_Color", Color(1,healthPercentage,healthPercentage,1));
	
}

function CheckBounds(param : float, min : float, max : float) {
	if (param < min) {
		param = min;
	} else if (param > max) {
		param = max;
	}
	
	return param;
}

File: Assets/scripts/c src/DamageableComponent.cs
using UnityEngine;
using System.Collections;

public class DamageableComponent : MonoBehaviour {

	public float health = 100.0f;
	public float maxHealth = 100.0f;
	public int mazeDifficulty = 7;
	public int mazeSize = 5;

	private ResourceComponent resourceScript;

	// Use this for initialization
	void Start () {
		resourceScript = gameObject.GetComponent<ResourceComponent> ();
	}
	
	// Update is called once per frame
	void Update () {
	
	}

	void TakeDamage (float amount) {
		health -= amount;
		health = CheckBounds( 0.0f, maxHealth, health);
		SetResourceRate();
		SetColor(gameObject);
	}

	void SetHealth(float amount) {
		health = amount;
		
		health = CheckBounds( 0.0f, maxHealth, health);
		SetResourceRate();
		SetColor(gameObject);
	}
	
	void Repair (float amount) {
		health += amount;
		health = CheckBounds( 0.0f, maxHealth, health);
		SetResourceRate();
		SetColor(gameObject);
	}

	void FullRepair() {
		health = maxHealth;
		SetResourceRate();
		SetColor(gameObject);
	}

	void SetResourceRate() {
		if (resourceScript != null) {
			resourceScript.SetOutputPercent(health / maxHealth);
		}

	}

	public int GetMazeDifficulty() {
		return mazeDifficulty;
	}

	public int GetMazeSize() {
		return mazeSize;
	}

	void SetColor(GameObject targetObject) {
		float healthPercentage = health / maxHealth;

		if (targetObject.GetComponent<Renderer>() == null) {
			// then game object is an empty parent, and the color of all of the children needs to be set.
			foreach (Transform child in targetObject.transform) {
				SetColor(child.gameObject);
			}
		} else {
			targetObject.GetComponent<Renderer>().material.SetColor("_Color", new Color(1,healthPercentage,healthPercentage,1));
		}
	}

	float CheckBounds(float min, float max, float param) {
		if (param < min) {
			param = min;
		} else if (param > max) {
			param = max;
		}
		
		return param;
	}

}
